### DISCUSSION TITLE
Counters for inputs coming in via PCF8574, ESP32

### DISCUSSION THREAD

**Discussion by hagenbuch** [arendst/Tasmota#21988] [Ideas]
I am planning to use a [Kincony KC868-AI device](https://www.kincony.com/kc868-ai-esp32-input-module.html) with Tasmota. It contains an ESP32, even with LAN / Ethernet.

Some other devices of Kincony are [already being supported](https://blakadder.com/kincony-KC868-A8/), even via blakadder's templates, which is awesome!

The KC868-AI offers 48 digital inputs, all are acquired via six PCF8574 Chips.

Firstly, I guess there should be no problem configuring Tasmota to use the inputs as mere inputs and read them, as Tasmota already offers USE_PCF8574_SENSOR (see [Blakadder's article](https://blakadder.com/kincony-KC868-A8/))

Now, I would like to use the inputs as counters, that is, each HI to LO transition (or LO to HI) should increment a counter which should then be somehow readable.

In the docs, I've read there is a counter feature already but it seems to work only for inputs wired directly to the ESP - or can I add the counter feature even for the inputs that are read in via I2C?

Having the idea that the chip could count on its own, I have browsed the [PCF8574 manual](https://www.ti.com/lit/ds/symlink/pcf8574.pdf) but it does not mention the word "counter" (except in a different context). Thus, I guess it will have to be a software counter...

I know it's not that easy because some debouncing should occur. Also, the counter values should survive a power fail, if possible - at least store some older status in flash.

As I am planning to use the counters for water counters that provide 1 pulse per litre, I guess 1 Hz or lower will be sufficient.

Thankful for every hint!
Andreas

**Comment by lalo-uy** [comment]
I belive tha it depends if the expanders interrupt pin is wired to the cpu.
Can you verify the circuit?

**Reply by barbudor** [reply]
Counters are strictly supported only on ESP's GPIOs
It's tighly tied to ESP interrupts

Porting to expanders would be very complicated

**Reply by hagenbuch** [reply]
Ok thanks! (If there'd be a polling of each byte every x msecs it might be doable though, without interrupt).

**Comment by lalo-uy** [comment]
If the frequency is low as you say, may be you can count with a button rule
;
“
Button2#State when a button changes state:
0 = OFF
1 = ON
2 = TOGGLE
3 = HOLD
“

**Reply by hagenbuch** [reply]
Yes thank you - but this would imply the use of 48 rules :)

BTW, as we are using ESP32 here - would the use of BerryScript make it possible to write (fast enough) code, like

- Every xx msec:
- Fetch 8 input bits from each PCF8574
- Iterate through each bit:
- Increment or reset debounce counters so a status is being changed only after the same logical value is present for n consecutive runs
- Increment this bit's counter if status goes from HI to LO

And then, every once in a while: Save counters to flash..

**Reply by sfromis** [reply]
With Berry, it would also be doable to handle the Button/Switch inputs without needing an "explosion" in number of `tasmota.add_rule`. Note that the normal restrictions on number of such inputs still applies.
